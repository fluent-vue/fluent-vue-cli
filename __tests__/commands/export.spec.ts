import { sync as rimraf } from 'rimraf'

import fs from 'fs'
import path from 'path'

import ExportCommand from '../../src/commands/export'

interface FileInfo {
  filepath: string
  content: string
}

/**
 * Read files synchronously from a folder, with natural sorting
 * @param dir Absolute path to directory
 * @returns File content
 */
function readFilesSync (dir: string): FileInfo[] {
  return fs.readdirSync(dir)
    .flatMap(filename => {
      const filepath = path.resolve(dir, filename)
      const stat = fs.statSync(filepath)
      const isFile = stat.isFile()

      if (isFile) {
        const content = fs.readFileSync(filepath).toString()

        return { filepath: path.relative(process.cwd(), filepath), content }
      }

      return readFilesSync(filepath)
    })
}

describe('export', () => {
  let result: string | Uint8Array | undefined

  beforeEach(() => {
    result = undefined
    jest
      .spyOn(process.stdout, 'write')
      .mockImplementation((val: string | Uint8Array): boolean => {
        result = val
        return true
      })
  })

  afterEach(() => {
    jest.restoreAllMocks()

    rimraf('translations')
  })

  it('extracts ftl messages from .vue files', async () => {
    // Act
    await ExportCommand.run(['__tests__/fixtures/*.vue'])

    // Assert
    expect(result).toContain('Extracted messages from 2 files')

    readFilesSync('translations').forEach(file => {
      expect(file.content).toMatchSnapshot(file.filepath)
    })
  })
})
