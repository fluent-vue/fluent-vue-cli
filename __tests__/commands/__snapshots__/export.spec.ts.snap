// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`export extracts ftl messages from .vue files: translations/en/__tests__/fixtures/Multiple.vue.ftl 1`] = `
"user-name = World
aria-key = Aria value
greeting = Hello, {$name}
  .aria-label = Label value
"
`;

exports[`export extracts ftl messages from .vue files: translations/en/__tests__/fixtures/Simple.vue.ftl 1`] = `
"user-name = World
aria-key = Aria value
greeting = Hello, {$name}
  .aria-label = Label value
"
`;

exports[`export extracts ftl messages from .vue files: translations/uk/__tests__/fixtures/Multiple.vue.ftl 1`] = `
"user-name = Світ
aria-key = Значення aria
greeting = Привіт, {$name}
  .aria-label = Значення мітки
"
`;
